versioning/worker-versions.md

```diff
@@ -0,0 +1,191 @@
+users will explicitly indicate that an ID is compatible with some already-extant
+ID.
+
+Because this behavior is a breaking change, SDKs will need to add a
+`enableBuildIdBasedVersioning` boolean flag to their `WorkerOptions`.
+
+## Server-side APIs for `workerBuildId`-based versioning
+
+Importantly, we _already_ send the id (checksum, today) on every poll request.
+It's TBD if the worker will be changed to poll a unique queue per ID, or if
+that will happen on the frontend, or matching.
+
+These APIs could very well end up in a different service, but are shown
+as part of `WorkflowService` here for simplicity.
+```protobuf
+syntax = "proto3";
+
+service WorkflowService {
+    rpc SetWorkerBuildIdOrdering (SetWorkerBuildIdOrderingRequest) returns (SetWorkerBuildIdOrderingResponse) {}
+    // This could / maybe should just be part of `DescribeTaskQueue`, but is broken out here to show easily.
+    rpc GetWorkerBuildIdOrdering (GetWorkerBuildIdOrderingRequest) returns (GetWorkerBuildIdOrderingResponse) {}
+}
+
+message SetWorkerBuildIdOrderingRequest {
+    string namespace = 1;
+    // Must be set, the task queue to apply changes to.
+    string task_queue = 2;
+    // The version id we are targeting.
+    VersionId version_id = 3;
+    // When set, indicates that the `version_id` in this message is compatible
+    // with the one specified in this field. Because compatability should form
+    // a DAG, any build id can only be the "next compatible" version for one
+    // other ID of a certain type at a time, and any setting which would create a cycle is invalid.
+    VersionId is_compatible_with = 4;
+    // When set, establishes the specified `version_id` as the default of it's type
+    // for the queue. Workers matching it will begin processing new workflow executions.
+    // The existing default will be marked as a previous incompatible version
+    // to this one, assuming it is not also in `is_compatible_with`.
+    bool default = 5;
+}
+message SetWorkerBuildIdOrderingResponse {}
+
+message GetWorkerBuildIdOrderingRequest {
+    string namespace = 1;
+    // Must be set, the task queue to interrogate about worker id ordering
+    string task_queue = 2;
+    // Limits how deep the returned DAG will go. 1 will return only the
+    // default build id.
+    uint32 max_depth = 3;
+}
+message GetWorkerBuildIdOrderingResponse {
+    // The currently established default ids. There will be one per version type as defined in
+    // the `VersionId` message.
+    repeated VersionIdNode current_defaults = 1;
+    // Other current latest-compatible versions who are not the overall default. These are the
+    // versions that will be used when generating new tasks by following the graph from the
+    // version of the last task out to a leaf.
+    repeated VersionIdNode compatible_leaves = 2;
+}
+
+message VersionIdNode {
+    VersionId version = 1;
+    VersionIdNode compatible_with = 2;
```

Comment:
I believe there are some limitation on recursion in proto. Need to double check.

Reply:
It can't have cycles, but we already explicitly will enforce that.

Reply:
Even w/o cycles there is performance degradation with deep depth.

Reply:
Yeah, any large depth is going to be a no-no here too. The perf characteristics will be awful. We will explicitly limit the depth which I mentioned in an easily-missed comment towards the bottom.